services/worker-manager/src/providers/azure.js

```
@@ -0,0 +1,379 @@
+    this.monitor.log.workerRunning({
+      workerPoolId: workerPool.workerPoolId,
+      providerId: this.providerId,
+      workerId: worker.workerId,
+    });
+    await worker.modify(w => {
+      w.lastModified = new Date();
+      w.state = this.Worker.states.RUNNING;
+    });
+    // assume for the moment that workers self-terminate before 96 hours
+    return {expires: taskcluster.fromNow('96 hours')};
+  }
+
+  async scanPrepare() {
+    this.seen = {};
+    this.errors = {};
+  }
+
+  async checkWorker({worker}) {
+    const states = this.Worker.states;
+    this.seen[worker.workerPoolId] = this.seen[worker.workerPoolId] || 0;
+    this.errors[worker.workerPoolId] = this.errors[worker.workerPoolId] || [];
+    if (worker.providerData.registrationExpiry &&
+      worker.state === states.REQUESTED &&
+      worker.providerData.registrationExpiry < Date.now()) {
+      return await this.removeWorker({worker});
+    }
+
+    let state = worker.state;
+    try {
+      const {provisioningState} = await this._enqueue('get', () => this.computeClient.virtualMachines.get(
+        worker.providerData.resourceGroupName,
+        worker.providerData.vm.name,
+      ));
+      if (['Creating', 'Updating', 'Starting', 'Running', 'Succeeded'].includes(provisioningState)) {
+        this.seen[worker.workerPoolId] += worker.capacity || 1;
+
+        // If the worker will be expired soon but it still exists,
+        // update it to stick around a while longer. If this doesn't happen,
+        // long-lived instances become orphaned from the provider. We don't update
+        // this on every loop just to avoid the extra work when not needed
+        if (worker.expires < taskcluster.fromNow('1 day')) {
+          await worker.modify(w => {
+            w.expires = taskcluster.fromNow('1 week');
+          });
+        }
+      } else if (['Failed', 'Canceled', 'Deleting', 'Deallocating', 'Stopped', 'Deallocated'].includes(provisioningState)) {
```

Are there more states than this or the possibility of them adding more? We could maybe at least report an error on unknown states